part4.md

```diff
@@ -0,0 +1,139 @@
+``` kotlin
+fun main(args: Array<String>) {
+    Admin.initializeApp(Functions.config().firebase)
+
+    exports.testPush = Functions.https.onRequest { req, res ->
+        Admin.database().ref("/conversions")
+                .push(InputConversions(req.query.eur.toString().toFloat()))
+                .then { res.status(200).send("done") }
+    }
+}
+```
+
+ * Build, deploy and make sure it still works.
+
+ * Copy the rest of my custom bindings (`bindings` folder in this repo) into `src/main/kotlin/bindings` 
+
+ * **Optional:** Examine copied bindings to see how it represents respective JavaScript APIs.
+
+ * Rewrite database table types to Kotlin:
+``` kotlin
+data class Rates(
+    val usd: Float,
+    val czk: Float,
+    val pln: Float,
+    val rub: Float
+)
+
+data class Conversions(
+    val eur: Float,
+    val usd: Float? = null,
+    val czk: Float? = null,
+    val pln: Float? = null,
+    val rub: Float? = null,
+    val timestamp: Long? = null
+)
+```
+
+ * Rewrite `convert` function to Kotlin:
+``` kotlin
+exports.convert = Functions.database.ref("/conversions/{conversionID}").onWrite<Conversions> { event ->
+    val input = event.data.`val`()
+    if (input.timestamp != null) return@onWrite
+    Admin.database().ref("/rates").once<Rates>("value")
+        .then<Conversions> { ratesSnap ->
+            ratesSnap.`val`().let {
+                Conversions(
+                    eur = input.eur,
+                    usd = input.eur * it.usd,
+                    czk = input.eur * it.czk,
+                    pln = input.eur * it.pln,
+                    rub = input.eur * it.rub,
+                    timestamp = Date.now()
+                )
+            }
+        }
+        .then { result ->
+            Admin.database().ref("/conversions")
+                .child(event.params.conversionID)
+                .set(result)
+        }
+}
+```
+
+ * Build and deploy to verify that it still works.
+
+### And a bit more Kotlin
+
+ * Move data classes to separate package e.g. `domain`
+
+ * In `Conversions.kt` add functions `inNotNew` and `convert(Rates, Long)`:
+``` kotlin
+fun Conversions.isNotNew() = timestamp != null
+
+fun Conversions.convert(rates: Rates, timestamp: Long) = Conversions(
+    eur = eur,
+    usd = eur * rates.usd,
+    czk = eur * rates.czk,
+    pln = eur * rates.pln,
+    rub = eur * rates.rub,
+    timestamp = timestamp
+)
+```
+
+ * Make use of new `Conversions` code:
+``` kotlin
+fun main(args: Array<String>) {
+    Admin.initializeApp(Functions.config().firebase)
+    val database = Admin.database()
+
+    exports.testPush = Functions.https.onRequest { req, res ->
+        database.ref("/conversions")
+            .push(Conversions(req.query.eur.toString().toFloat()))
+            .then { res.status(200).send("done") }
+    }
+
+    exports.convert = Functions.database.ref("/conversions/{conversionID}").onWrite<Conversions> { event ->
+        event.data.`val`().takeUnless { it.isNotNew() }?.let { input ->
```

Comment:
takeUnless it is not new - double negative 😆 
make the fun `isNew` and use `takeIf` ?

Reply:
Literally no difference to me :D